source/common/common/bit_array.h

```diff
@@ -0,0 +1,125 @@
+    const size_t index_of_0th_bit = bit0_offset & 0x7;
+    // It is not in the last word, we can memcpy.
+    if (byte0 + WordSize <= end_) {
+      return (LoadUnsignedWord(byte0) >> index_of_0th_bit) & mask_;
+    }
+    // We're at the last word, and we've allocated at least a word of memory.
+    const uint8_t* last_word = end_ - WordSize;
+    return LoadUnsignedWord(last_word) >> (index_of_0th_bit + ((byte0 - last_word) << 3)) & mask_;
+  }
+
+  /**
+   * Sets the given fixed width bit to the given value.
+   */
+  inline void set(size_t index, uint32_t value) {
+    if (index > num_items_) {
+      ENVOY_BUG(true, "BitArray::set requested index out of bounds");
+      return;
+    }
+    const size_t bit0_offset = index * bit_width_;
+    const size_t byte0_offset = bit0_offset >> 3;
+    uint8_t* byte0 = array_start_.get() + byte0_offset;
+
+    // We are not at the end.
+    if (byte0 + WordSize <= end_) {
+      const size_t index_of_0th_bit = bit0_offset & 0x7;
+      const uint64_t shifted_value = static_cast<uint64_t>(value) << index_of_0th_bit;
+
+      const uint64_t mask_to_write = ((static_cast<uint64_t>(1) << bit_width_) - 1)
+                                     << index_of_0th_bit;
+      const uint64_t mask_to_preserve = ~mask_to_write;
+      const uint64_t value_to_store =
+          ((LoadUnsignedWord(byte0) & mask_to_preserve) | (shifted_value & mask_to_write));
+      StoreUnsignedWord(byte0, value_to_store);
+    } else {
+      // We're at the end, so we'll be starting from the last word
+      uint8_t* last_word = const_cast<uint8_t*>(end_) - WordSize;
+      const size_t index_of_0th_bit = (bit0_offset & 0x7) + ((byte0 - last_word) << 3);
+      const uint64_t shifted_value = static_cast<uint64_t>(value) << index_of_0th_bit;
+      const uint64_t mask_to_write = ((static_cast<uint64_t>(1) << bit_width_) - 1)
+                                     << index_of_0th_bit;
+      const uint64_t mask_to_preserve = ~mask_to_write;
+      const uint64_t value_to_store =
+          ((LoadUnsignedWord(last_word) & mask_to_preserve) | (shifted_value & mask_to_write));
+      return StoreUnsignedWord(last_word, value_to_store);
+    }
+  }
+
+  size_t size() const { return num_items_; }
+
+private:
+  static inline size_t bytesNeeded(int bit_width, size_t num_items) {
+    // Round up number of bytes needed.
+    // To simplify the implementation when accessing the last elements,
+    // we must allocate a minimum of a word.
+    return std::max((bit_width * num_items + 7) >> 3, static_cast<size_t>(WordSize));
+  }
+
+  static inline void StoreUnsignedWord(void* destination, uint64_t value) {
+    safeMemcpyUnsafeDst(destination, &value);
+  }
+
+  static inline uint64_t LoadUnsignedWord(const void* source) {
+    uint64_t destination;
+    safeMemcpyUnsafeSrc(&destination, source);
+    return destination;
+  }
+
+  // Backing storage for the underlying array of bits.
+  std::unique_ptr<uint8_t[]> array_start_;
```

Comment:
Why not use `std::vector`, since it'll only be an additional 8 bytes?

Reply:
Makes sure I alloc the exact amount of memory without having to do any shrink to fit or similar dance. 

The vec for 10k might stl vec have size != capacity and alloc additional space.